GuessMyNumber/Main.lean

```diff
@@ -0,0 +1,27 @@
+import Std.Data
+
+-- Generate a "random" number in [0, 99]
+-- It's not actually random. We'll talk about it later.
+def GetSecret : IO Nat :=
+  IO.rand 0 99
+
+partial def guess (secret : Nat) (prompt : String) : IO Unit := do
+  IO.println prompt
+  let stdin := (←IO.getStdin)
```

Comment:
This is correct, but most Lean programmers would write
```lean
  let stdin ← IO.getStdin
```
The same pattern occurs in other places.

Reply:
let's show both forms here for additional educational value.

Reply:
There is no reason for using `let stdin := (←IO.getStdin)`, it is a bad style.

Reply:
Oh, interesting.  If it is "bad" why doesn't Lean remove it?  What bad things can happen?  Or are you just saying it is not the preferred way in a "lint" sort of "coding guidlines" kind of way?

Reply:
> Oh, interesting. If it is "bad" why doesn't Lean remove it?

It is not incorrect to write `let stdin := (←IO.getStdin)`. The program will behave as expected.

> Or are you just saying it is not the preferred way in a "lint" sort of "coding guidlines" kind of way?

Yes, this is what I mean by "bad style". In the future, someone may write a linter for reporting a warning when someone uses this style, but this is super low-priority. We should just make sure documentation and our examples do not use "bad style".

Reply:
Well in my updated version of this (see ) I'm making this an educational moment by writing:

```lean
let stdin := (←IO.getStdin)
let str ← stdin.getLine
```

The `←` notation is just to get things out of the IO box. The type of `stdin.getLine` is `IO
String`, and `str` is of type `String`. Notice the second let here uses a short form of assignment
without the need for the `:=` operator and this is the preferred way of doing it.


Is this ok?

Reply:
I don't like it.  Note that `let str ← stdin.getLine` is **not** a syntax sugar for `let str := (← stdin.getLine)`. The first one is the "macro". 
```lean
let stdin := (←IO.getStdin)
```
is expanded to
```lean
let aux ← IO.getStdin
let stdin := aux
```
I think it is better to just write code as everybody does it, and let David's book explain monadic code and nested actions.